examples/simple/simple-backend.cpp

```diff
@@ -29,111 +17,90 @@ static void ggml_log_callback_default(ggml_log_level level, const char * text, v
 
-    // initialize the backend
-#ifdef GGML_USE_CUDA
-    fprintf(stderr, "%s: using CUDA backend\n", __func__);
-    model.backend = ggml_backend_cuda_init(0); // init device 0
-    if (!model.backend) {
-        fprintf(stderr, "%s: ggml_backend_cuda_init() failed\n", __func__);
-    }
-#endif
+    ggml_backend_load_all();
 
-#ifdef GGML_USE_METAL
-    fprintf(stderr, "%s: using Metal backend\n", __func__);
-    model.backend = ggml_backend_metal_init();
-    if (!model.backend) {
-        fprintf(stderr, "%s: ggml_backend_metal_init() failed\n", __func__);
-    }
-#endif
+    model.backend = ggml_backend_init_best();
+    model.cpu_backend = ggml_backend_init_by_type(GGML_BACKEND_DEVICE_TYPE_CPU, nullptr);
 
-    // if there aren't GPU Backends fallback to CPU backend
-    if (!model.backend) {
-        model.backend = ggml_backend_cpu_init();
-    }
+    ggml_backend_t backends[2] = { model.backend, model.cpu_backend };
+    model.sched = ggml_backend_sched_new(backends, nullptr, 2, GGML_DEFAULT_GRAPH_SIZE, false, true);
 
-    int num_tensors = 2;
+    size_t buf_size = ggml_tensor_overhead()*GGML_DEFAULT_GRAPH_SIZE + ggml_graph_overhead();
 
-    struct ggml_init_params params {
-            /*.mem_size   =*/ ggml_tensor_overhead() * num_tensors,
-            /*.mem_buffer =*/ NULL,
-            /*.no_alloc   =*/ true,
+    struct ggml_init_params params0 = {
+        /*.mem_size   =*/ buf_size,
+        /*.mem_buffer =*/ nullptr,
+        /*.no_alloc   =*/ true, // the tensors will be allocated later
     };
 
-    // create context
-    model.ctx = ggml_init(params);
-
-    // create tensors
-    model.a = ggml_new_tensor_2d(model.ctx, GGML_TYPE_F32, cols_A, rows_A);
-    model.b = ggml_new_tensor_2d(model.ctx, GGML_TYPE_F32, cols_B, rows_B);
-
-    // create a backend buffer (backend memory) and alloc the tensors from the context
-    model.buffer = ggml_backend_alloc_ctx_tensors(model.ctx, model.backend);
-
-    // load data from cpu memory to backend buffer
-    ggml_backend_tensor_set(model.a, a, 0, ggml_nbytes(model.a));
-    ggml_backend_tensor_set(model.b, b, 0, ggml_nbytes(model.b));
+    // create a context to build the graph
+    model.ctx = ggml_init(params0);
 }
 
 // build the compute graph to perform a matrix multiplication
-struct ggml_cgraph * build_graph(const simple_model& model) {
-    static size_t buf_size = ggml_tensor_overhead()*GGML_DEFAULT_GRAPH_SIZE + ggml_graph_overhead();
-    static std::vector<uint8_t> buf(buf_size);
-
-    struct ggml_init_params params0 = {
-        /*.mem_size   =*/ buf_size,
-        /*.mem_buffer =*/ buf.data(),
-        /*.no_alloc   =*/ true, // the tensors will be allocated later by ggml_gallocr_alloc_graph()
-    };
-
-    // create a temporally context to build the graph
-    struct ggml_context * ctx0 = ggml_init(params0);
+struct ggml_cgraph * build_graph(simple_model& model) {
+    struct ggml_cgraph  * gf = ggml_new_graph(model.ctx);
```

Comment:
While not strictly necessary for this example since the graph is only built once, applications usually use a different `ggml_context` for the graph that is reset every time a new graph is created (effectively freeing all the tensors allocated for the previous graph). It may be a better example if this separation between model context and graph context is maintained here.